substratevm/src/com.oracle.svm.hosted.jdk11/src/com/oracle/svm/hosted/ModuleLayerFeature.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (c) 2021, 2021, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.oracle.svm.hosted;
+
+import com.oracle.graal.pointsto.meta.AnalysisUniverse;
+import com.oracle.svm.core.annotate.AutomaticFeature;
+import com.oracle.svm.core.jdk.BootModuleLayerSupport;
+import com.oracle.svm.core.jdk.JDK11OrLater;
+import org.graalvm.nativeimage.ImageSingletons;
+import org.graalvm.nativeimage.Platform;
+import org.graalvm.nativeimage.Platforms;
+import org.graalvm.nativeimage.hosted.Feature;
+
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@AutomaticFeature
+@Platforms(Platform.HOSTED_ONLY.class)
+public final class ModuleLayerFeature implements Feature {
+
+    @Override
+    public boolean isInConfiguration(IsInConfigurationAccess access) {
+        return new JDK11OrLater().getAsBoolean();
+    }
+
+    @Override
+    public void beforeAnalysis(BeforeAnalysisAccess access) {
+        ImageSingletons.add(BootModuleLayerSupport.class, new BootModuleLayerSupport());
+    }
+
+    @Override
+    public void afterAnalysis(AfterAnalysisAccess access) {
+        FeatureImpl.AfterAnalysisAccessImpl accessImpl = (FeatureImpl.AfterAnalysisAccessImpl) access;
+        AnalysisUniverse universe = accessImpl.getUniverse();
+
+        Set<Object> moduleSet = universe.getTypes()
+                .stream()
+                .filter(t -> t.isReachable() && !t.isArray())
+                .map(t -> t.getJavaClass().getModule())
+                .filter(Module::isNamed)
```

Comment:
Rather than filtering out the unnamed module we should understand why we have it in the analysis results if we built a pure module-path based image. Pleas investigate what causes it to show up in the module-path based image usecase.

Reply:
After investigating this a bit more, I discovered that in my case after analysis there are around 135 reachable classes in the unnamed module. All o135 classes seem to be synthesized from class methods contained in 
`com.oracle.core.svm.reflect`. 

I will keep updating this comment as I investigate this further.

Reply:
Do we already know what was causing methods from `com.oracle.core.svm.reflect` to have ties to the unnamed module?

Reply:
Initially I thought that these classes are only related to `VarHandle`s but I was wrong. From what I can understand, they are generated for the runtime method reflection support - so that `Class.getMethod` will work properly. Now I am not sure if these generated classes belong in the unnamed module, I assume that is the case?

Reply:
I suspect these generated classes for runtime method reflection should be in the same module as the classes are they are generated for. This needs to be fixed in the runtime method reflection support.

Reply:
But in any case, the unnamed module should not be part of the boot layer, right? I am assuming it would be the best to keep this filter, just to be safe?

Reply:
> But in any case, the unnamed module should not be part of the boot layer, right?

This is correct. To verify that it really is I created an example where a java application uses classes from the module path and the classpath at the same time (using an automatic-module as a bridge from modulepath to classpath) and then dumps the contents of `ModuleLayer.boot()`. As expected the involved unnamed module does **not** show up in `ModuleLayer.boot()`.
```
$ java -cp cp -p mp:mp-to-cp.jar -m trivialmodule/trivialpackage.HelloJava
Hello 𝓙𝓪𝓿𝓪 ☕
Hello from HelloJava class in module trivialmodule
Hello from ClassLoadedFromAutomaticModule class in module mp.to.cp
Hello From Class on Classpath unnamed module @7ac7a4e4
Modules in the boot layer:
module jdk.crypto.ec
module java.rmi
module java.management.rmi
module java.logging
module jdk.localedata
module jdk.security.jgss
module jdk.naming.rmi
module java.xml
module java.datatransfer
module java.desktop
module trivialmodule
module java.naming
module jdk.zipfs
module jdk.management.jfr
module jdk.jdeps
module java.xml.crypto
module jdk.charsets
module java.smartcardio
module java.prefs
module jdk.jfr
module java.management
module jdk.javadoc
module jdk.compiler
module java.security.jgss
module jdk.internal.opt
module jdk.unsupported.desktop
module java.compiler
module jdk.jartool
module jdk.crypto.cryptoki
module mp.to.cp
module jdk.security.auth
module jdk.naming.dns
module java.base
module jdk.management
module java.security.sasl
module jdk.jlink
```
I added the stuff i hacked together as attachment: [test-mix-mp-cp.zip](https://github.com/oracle/graal/files/6739944/test-mix-mp-cp.zip)

In other words, yes please keep the filter. It's actually required.

That said, we still have to address the issue of creating generated classes for runtime method reflection to be in the same module as the classes they are generated for.